In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report, balanced_accuracy_score

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending = pd.read_csv(
    Path('Resources/lending_data.csv')
)
# Review the DataFrame
print(df_lending.shape)
df_lending.head()

(77536, 8)


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending["loan_status"]

# Separate the X variable, the features
X = df_lending.drop("loan_status", axis=1)

In [4]:
# Review the y variable Series
y[:10]

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X[:10]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
5                 1       20600  
6                 1       21100  
7                 0       15100  
8                 0       17400  
9                 0       18800

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(random_state=1, solver='lbfgs', max_iter=200)

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test}).tail(10)

Prediction  Actual
73999           0       0
47267           0       0
35950           0       0
42373           0       0
38631           0       0
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [13]:
# Generate a confusion matrix for the model
confuse_matrix = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(confuse_matrix,
                      index=["Actual 0", "Actual 1"],
                      columns=["Predicted 0", "Predicted 1"])

cm_df

Predicted 0  Predicted 1
Actual 0        18658          107
Actual 1           37          582

In [14]:
# Print the classification report for the model
target_names = ["(0) Healthy Loans","(1) High-Risk Loans"]
print(classification_report(y_test,predictions,target_names=target_names))

                     precision    recall  f1-score   support

  (0) Healthy Loans       1.00      0.99      1.00     18765
(1) High-Risk Loans       0.84      0.94      0.89       619

           accuracy                           0.99     19384
          macro avg       0.92      0.97      0.94     19384
       weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model has an impressive overall accuracy of 99%, indicating that it correctly classifies the vast majority of instances. However, there is a gap in the models ability to predict high-risk loans (1), with about 10% of high-risk loans being classified as healthy. This is possibly due to having significantly less data points of high-risk loans to use for training.
The macro average metrics (precision: 0.92, recall: 0.97, F1-score: 0.94) indicate good overall performance, treating both classes equally.
The weighted averages (precision: 0.99, recall: 0.99, F1-score: 0.99) reflect the model's performance considering the class distribution, showing that it performs exceptionally well overall, particularly for healthy loans.
In summary, the model is strong in predicting healthy loans and reasonably good in predicting high-risk loans, but there’s potential for improvement in the latter.


---